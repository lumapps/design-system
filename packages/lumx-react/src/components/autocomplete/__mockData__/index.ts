export const CITIES = [
    {
        text: 'Los Angeles',
        id: 'losangeles',
    },
    {
        text: 'San Francisco',
        id: 'sanfrancisco',
    },
    {
        text: 'Paris',
        id: 'paris',
    },
    {
        text: 'Montpellier',
        id: 'montpellier',
    },
    {
        text: 'Bordeaux',
        id: 'bordeaux',
    },
    {
        text: 'Toulouse',
        id: 'toulouse',
    },
    {
        text: 'Lyon',
        id: 'lyon',
    },
    {
        text: 'Montevideo',
        id: 'montevideo',
    },
];
