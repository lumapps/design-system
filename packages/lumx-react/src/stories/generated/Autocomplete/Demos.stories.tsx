/**
 * File generated when storybook is started. Do not edit directly!
 */
export default { title: 'LumX components/autocomplete/Demos' };

export { App as Default } from './default';
export { App as Multiple } from './multiple';
