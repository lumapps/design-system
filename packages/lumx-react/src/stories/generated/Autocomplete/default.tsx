/home/gcornut/git/design-system/packages/site-demo/content/product/components/autocomplete/react/default.tsx