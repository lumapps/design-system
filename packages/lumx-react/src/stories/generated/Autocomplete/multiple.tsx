/home/gcornut/git/dup-design-system/packages/site-demo/content/product/components/autocomplete/react/multiple.tsx