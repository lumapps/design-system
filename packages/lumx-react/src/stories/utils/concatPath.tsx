export type PathPart = string | number;
export type Path = PathPart | Array<PathPart>;

/**
 * Concat flatten object path
 *
 * @example concatPath('foo', 'bar') // => 'foo.bar'
 * @example concatPath(['foo', 0]) // => 'foo[0]'
 * @example concatPath('foo', 0, ['bar']) // => 'foo[0].bar'
 */
export const concatPath = (...prefix: Path[]) => {
    const [first, ...rest] = prefix.flat();
    return rest.reduce<string>((acc, part) => {
        if (typeof part === 'number') return `${acc}[${part}]`;
        return `${acc}.${part}`;
    }, String(first));
};
