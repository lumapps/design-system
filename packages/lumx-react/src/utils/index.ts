/**
 * WARNING: All modules exported here are exposed to NPM in '@lumx/react/utils'.
 */

export { ClickAwayProvider } from './ClickAwayProvider';
export { Portal, type PortalProps, type PortalInit, PortalProvider, type PortalProviderProps } from './Portal';
