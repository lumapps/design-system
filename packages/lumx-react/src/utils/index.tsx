export * from '@lumx/core/js/utils';

export * from './getRootClassName';

export * from './htmlDecode';

export * from './partitionMulti';

export * from './type';
