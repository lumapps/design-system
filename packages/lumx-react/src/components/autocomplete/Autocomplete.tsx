import React, { forwardRef, ReactNode, RefObject, SyntheticEvent, useRef } from 'react';

import classNames from 'classnames';

import { Dropdown, IconButtonProps, Offset, Placement, TextField, Theme } from '@lumx/react';

import { Comp, GenericProps, getRootClassName, handleBasicClasses } from '@lumx/react/utils';

import { useFocus } from '@lumx/react/hooks/useFocus';
import { mergeRefs } from '@lumx/react/utils/mergeRefs';

/**
 * Defines the props of the component.
 */
export interface AutocompleteProps extends GenericProps {
    /**
     * Whether the suggestions list should display anchored to the input or to the wrapper.
     * @see {@link DropdownProps#anchorToInput}
     */
    anchorToInput?: boolean;
    /**
     * Props to pass to the clear button (minus those already set by the TextField props).
     * If not specified, the button won't be displayed.
     * @see {@link TextFieldProps#clearButtonProps}
     */
    clearButtonProps?: Pick<IconButtonProps, 'label'> &
        Omit<IconButtonProps, 'label' | 'onClick' | 'icon' | 'emphasis'>;
    /**
     * Reference to the <input> or <textarea> element.
     * @see {@link TextFieldProps#inputRef}
     */
    inputRef?: RefObject<HTMLInputElement>;
    /**
     * The offset that will be applied to the Dropdown position.
     * @see {@link DropdownProps#offset}
     */
    offset?: Offset;
    /**
     * The preferred Dropdown location against the anchor element.
     * @see {@link DropdownProps#placement}
     */
    placement?: Placement;
    /**
     * Whether the dropdown should fit to the anchor width or not.
     * @see {@link DropdownProps#fitToAnchorWidth}
     */
    fitToAnchorWidth?: boolean;
    /**
     * The error related to the component.
     * @see {@link TextFieldProps#error}
     */
    error?: string | ReactNode;
    /**
     * Whether the text field is displayed with error style or not.
     * @see {@link TextFieldProps#hasError}
     */
    hasError?: boolean;
    /**
     * Whether the text box should be focused upon closing the suggestions or not.
     */
    shouldFocusOnClose?: boolean;
    /**
     * The helper message of the text field.
     * @see {@link TextFieldProps#helper}
     */
    helper?: string;
    /**
     * The icon of the text field (SVG path).
     * @see {@link TextFieldProps#icon}
     */
    icon?: string;
    /**
     * Whether the component is disabled or not.
     * @see {@link TextFieldProps#isDisabled}
     */
    isDisabled?: boolean;
    /**
     * Whether the component is required or not.
     * @see {@link TextFieldProps#isRequired}
     */
    isRequired?: boolean;
    /**
     * Whether the text field is displayed with valid style or not.
     * @see {@link TextFieldProps#isValid}
     */
    isValid?: boolean;
    /**
     * The label of the text field displayed in a label tag.
     * @see {@link TextFieldProps#label}
     */
    label?: string;
    /**
     * The placeholder message of the text field.
     * @see {@link TextFieldProps#placeholder}
     */
    placeholder?: string;
    /** Theme adapting the component to light or dark background. */
    theme?: Theme;
    /** List of suggestions to display during autocomplete. */
    children: React.ReactNode;
    /**
     * The list of chips to be displayed before the text field input.
     */
    chips?: React.ReactNode;
    /**
     * The value of the text field.
     * @see {@link TextFieldProps#value}
     */
    value: string;
    /**
     * Whether the suggestions from the autocomplete should be displayed or not.
     * @see {@link DropdownProps#isOpen}
     */
    isOpen: boolean;
    /**
     * The native input name property.
     * @see {@link TextFieldProps#name}
     */
    name?: string;
    /**
     * Whether a click in the Autocomplete dropdown would close it or not.
     * @see {@link DropdownProps#closeOnClick}
     */
    closeOnClick?: boolean;
    /**
     * Whether a click anywhere out of the Autocomplete would close it or not.
     * @see {@link DropdownProps#closeOnClickAway}
     */
    closeOnClickAway?: boolean;
    /**
     * Whether an escape key press would close the Autocomplete or not.
     * @see {@link DropdownProps#closeOnEscape}
     */
    closeOnEscape?: boolean;
    /**
     * The function called on blur.
     * @see {@link TextFieldProps#onBlur}
     */
    onBlur?(event: React.FocusEvent): void;
    /**
     * On change callback.
     * @see {@link TextFieldProps#onChange}
     */
    onChange(value: string, name?: string, event?: SyntheticEvent): void;
    /**
     * The function called on close.
     * @see {@link DropdownProps#onClose}
     */
    onClose?(): void;
    /**
     * The function called on focus.
     * @see {@link TextFieldProps#onFocus}
     */
    onFocus?(event: React.FocusEvent): void;
    /**
     * The function called when the bottom of the dropdown is reached.
     * @see {@link DropdownProps#onInfiniteScroll}
     */
    onInfiniteScroll?(): void;
}

/**
 * Component display name.
 */
const COMPONENT_NAME = 'Autocomplete';

/**
 * Component default class name and class prefix.
 */
const CLASSNAME = getRootClassName(COMPONENT_NAME);

/**
 * Component default props.
 */
const DEFAULT_PROPS: Partial<AutocompleteProps> = {
    anchorToInput: false,
    closeOnClick: false,
    closeOnClickAway: true,
    closeOnEscape: true,
    shouldFocusOnClose: false,
};

/**
 * Autocomplete component.
 *
 * @param  props Component props.
 * @param  ref   Component ref.
 * @return React element.
 */
export const Autocomplete: Comp<AutocompleteProps, HTMLDivElement> = forwardRef((props, ref) => {
    const {
        anchorToInput,
        children,
        chips,
        className,
        closeOnClick,
        closeOnClickAway,
        closeOnEscape,
        disabled,
        error,
        fitToAnchorWidth,
        hasError,
        helper,
        icon,
        inputRef,
        clearButtonProps,
        isDisabled = disabled,
        isRequired,
        isOpen,
        isValid,
        label,
        name,
        offset,
        onBlur,
        onChange,
        onClose,
        onFocus,
        onInfiniteScroll,
        placeholder,
        placement,
        shouldFocusOnClose,
        theme,
        value,
        ...forwardedProps
    } = props;
    const inputAnchorRef = useRef<HTMLElement>(null);
    const textFieldRef = useRef(null);
    useFocus(inputAnchorRef.current, !isOpen && shouldFocusOnClose);

    return (
        <div
            ref={ref}
            {...forwardedProps}
            className={classNames(
                className,
                handleBasicClasses({
                    prefix: CLASSNAME,
                }),
            )}
        >
            <TextField
                chips={chips}
                error={error}
                hasError={hasError}
                helper={helper}
                icon={icon}
                inputRef={mergeRefs(inputAnchorRef, inputRef) as any}
                clearButtonProps={clearButtonProps}
                isDisabled={isDisabled}
                isRequired={isRequired}
                isValid={isValid}
                label={label}
                name={name}
                onBlur={onBlur}
                onChange={onChange}
                onFocus={onFocus}
                placeholder={placeholder}
                textFieldRef={textFieldRef}
                theme={theme}
                value={value}
            />
            <Dropdown
                anchorRef={anchorToInput ? inputAnchorRef : textFieldRef}
                closeOnClick={closeOnClick}
                closeOnClickAway={closeOnClickAway}
                closeOnEscape={closeOnEscape}
                fitToAnchorWidth={fitToAnchorWidth}
                isOpen={isOpen}
                offset={offset}
                onClose={onClose}
                onInfiniteScroll={onInfiniteScroll}
                placement={placement}
                shouldFocusOnOpen={false}
                theme={theme}
            >
                {children}
            </Dropdown>
        </div>
    );
});
Autocomplete.displayName = COMPONENT_NAME;
Autocomplete.className = CLASSNAME;
Autocomplete.defaultProps = DEFAULT_PROPS;
