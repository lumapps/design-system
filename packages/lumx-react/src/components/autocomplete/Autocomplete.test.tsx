import React, { ReactElement } from 'react';

import { mount, shallow } from 'enzyme';
import 'jest-enzyme';

import { List, ListItem, Size } from '@lumx/react';
import { commonTestsSuite, Wrapper } from '@lumx/react/testing/utils';
import { Autocomplete, AutocompleteProps } from './Autocomplete';

import { CITIES as suggestions } from './__mockData__';

const CLASSNAME = Autocomplete.className as string;

type SetupProps = Partial<AutocompleteProps>;

interface Suggestion {
    id: string;
    text: string;
}

/**
 * Mounts the component and returns common DOM elements / data needed in multiple tests further down.
 */
const setup = (propsOverride: SetupProps = {}, shallowRendering = true) => {
    const props: any = { ...propsOverride };
    const renderer: (el: ReactElement) => Wrapper = shallowRendering ? shallow : mount;

    const wrapper: Wrapper = renderer(<Autocomplete {...props} />);

    const textField: Wrapper = wrapper.find('TextField');
    const dropdown: Wrapper = wrapper.find('Dropdown');

    return { dropdown, props, textField, wrapper };
};

describe(`<${Autocomplete.displayName}>`, () => {
    // 1. Test render via snapshot (default states of component).
    describe('Snapshots and structure', () => {
        // Here is an example of a basic rendering check, with snapshot.

        it('should render correctly', () => {
            const { wrapper, textField, dropdown } = setup({
                children: (
                    <List isClickable>
                        {suggestions.map((suggestion: Suggestion) => (
                            <ListItem size={Size.tiny} key={suggestion.id}>
                                <div>{suggestion.text}</div>
                            </ListItem>
                        ))}
                    </List>
                ),
                label: 'Field label',
                isOpen: true,
                isRequired: true,
                onChange: jest.fn(),
                value: '',
            });
            expect(wrapper).toMatchSnapshot();

            expect(wrapper).toExist();

            expect(wrapper).toHaveClassName(CLASSNAME);

            expect(textField).toHaveLength(1);
            expect(dropdown).toHaveLength(1);
        });
    });

    // 2. Test defaultProps value and important props custom values.
    describe('Props', () => {
        it('should use default props', () => {
            const { wrapper } = setup();

            expect(wrapper).toHaveClassName(CLASSNAME);
        });

        it('should render correctly when the dropdown is closed', () => {
            const { wrapper } = setup({
                children: (
                    <List isClickable>
                        {suggestions.map((suggestion: Suggestion) => (
                            <ListItem size={Size.tiny} key={suggestion.id}>
                                <div>{suggestion.text}</div>
                            </ListItem>
                        ))}
                    </List>
                ),
                isOpen: false,
                onChange: jest.fn(),
                value: '',
            });
            expect(wrapper).toMatchSnapshot();

            expect(wrapper).toExist();

            expect(wrapper).toHaveClassName(CLASSNAME);
        });
    });

    // 3. Test events.
    describe('Events', () => {
        it('should trigger the onChange callback when there is a change on the Text Field', () => {
            const onChange = jest.fn();
            const { textField } = setup({
                children: (
                    <List isClickable>
                        {suggestions.map((suggestion: Suggestion) => (
                            <ListItem size={Size.tiny} key={suggestion.id}>
                                <div>{suggestion.text}</div>
                            </ListItem>
                        ))}
                    </List>
                ),
                isOpen: false,
                onChange,
                value: '',
            });

            textField.simulate('change');

            expect(onChange).toHaveBeenCalled();
        });

        it('should trigger the onFocus callback when the text field is focused on', () => {
            const onFocus = jest.fn();
            const { textField } = setup({
                children: (
                    <List isClickable>
                        {suggestions.map((suggestion: Suggestion) => (
                            <ListItem size={Size.tiny} key={suggestion.id}>
                                <div>{suggestion.text}</div>
                            </ListItem>
                        ))}
                    </List>
                ),
                isOpen: false,
                onFocus,
                value: '',
            });

            textField.simulate('focus');

            expect(onFocus).toHaveBeenCalled();
        });

        it('should trigger the onBlur callback when the Text Field loses focus', () => {
            const onBlur = jest.fn();
            const { textField } = setup({
                children: (
                    <List isClickable>
                        {suggestions.map((suggestion: Suggestion) => (
                            <ListItem size={Size.tiny} key={suggestion.id}>
                                <div>{suggestion.text}</div>
                            </ListItem>
                        ))}
                    </List>
                ),
                isOpen: false,
                onBlur,
                value: '',
            });

            textField.simulate('blur');

            expect(onBlur).toHaveBeenCalled();
        });
    });

    // 4. Test conditions (i.e. things that display or not in the UI based on props).
    describe('Conditions', () => {
        // Nothing to do here.
    });

    // 5. Test state.
    describe('State', () => {
        // Nothing to do here.
    });

    // Common tests suite.
    commonTestsSuite(setup, { className: 'wrapper', prop: 'wrapper' }, { className: CLASSNAME });
});
