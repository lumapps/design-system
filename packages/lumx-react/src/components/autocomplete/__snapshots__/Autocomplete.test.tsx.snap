// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Autocomplete> Props should render correctly when the dropdown is closed 1`] = `
<div
  className="lumx-autocomplete"
>
  <TextField
    inputRef={[Function]}
    onChange={[MockFunction]}
    textFieldRef={
      Object {
        "current": null,
      }
    }
    theme="light"
    type="text"
    value=""
  />
  <Dropdown
    anchorRef={
      Object {
        "current": null,
      }
    }
    closeOnClick={false}
    closeOnClickAway={true}
    closeOnEscape={true}
    fitToAnchorWidth={true}
    fitWithinViewportHeight={true}
    isOpen={false}
    placement="bottom-start"
    shouldFocusOnOpen={false}
  >
    <List
      isClickable={true}
    >
      <ListItem
        key="losangeles"
        size="tiny"
      >
        <div>
          Los Angeles
        </div>
      </ListItem>
      <ListItem
        key="sanfrancisco"
        size="tiny"
      >
        <div>
          San Francisco
        </div>
      </ListItem>
      <ListItem
        key="paris"
        size="tiny"
      >
        <div>
          Paris
        </div>
      </ListItem>
      <ListItem
        key="montpellier"
        size="tiny"
      >
        <div>
          Montpellier
        </div>
      </ListItem>
      <ListItem
        key="bordeaux"
        size="tiny"
      >
        <div>
          Bordeaux
        </div>
      </ListItem>
      <ListItem
        key="toulouse"
        size="tiny"
      >
        <div>
          Toulouse
        </div>
      </ListItem>
      <ListItem
        key="lyon"
        size="tiny"
      >
        <div>
          Lyon
        </div>
      </ListItem>
      <ListItem
        key="montevideo"
        size="tiny"
      >
        <div>
          Montevideo
        </div>
      </ListItem>
    </List>
  </Dropdown>
</div>
`;

exports[`<Autocomplete> Snapshots and structure should render correctly 1`] = `
<div
  className="lumx-autocomplete"
>
  <TextField
    inputRef={[Function]}
    isRequired={true}
    label="Field label"
    onChange={[MockFunction]}
    textFieldRef={
      Object {
        "current": null,
      }
    }
    theme="light"
    type="text"
    value=""
  />
  <Dropdown
    anchorRef={
      Object {
        "current": null,
      }
    }
    closeOnClick={false}
    closeOnClickAway={true}
    closeOnEscape={true}
    fitToAnchorWidth={true}
    fitWithinViewportHeight={true}
    isOpen={true}
    placement="bottom-start"
    shouldFocusOnOpen={false}
  >
    <List
      isClickable={true}
    >
      <ListItem
        key="losangeles"
        size="tiny"
      >
        <div>
          Los Angeles
        </div>
      </ListItem>
      <ListItem
        key="sanfrancisco"
        size="tiny"
      >
        <div>
          San Francisco
        </div>
      </ListItem>
      <ListItem
        key="paris"
        size="tiny"
      >
        <div>
          Paris
        </div>
      </ListItem>
      <ListItem
        key="montpellier"
        size="tiny"
      >
        <div>
          Montpellier
        </div>
      </ListItem>
      <ListItem
        key="bordeaux"
        size="tiny"
      >
        <div>
          Bordeaux
        </div>
      </ListItem>
      <ListItem
        key="toulouse"
        size="tiny"
      >
        <div>
          Toulouse
        </div>
      </ListItem>
      <ListItem
        key="lyon"
        size="tiny"
      >
        <div>
          Lyon
        </div>
      </ListItem>
      <ListItem
        key="montevideo"
        size="tiny"
      >
        <div>
          Montevideo
        </div>
      </ListItem>
    </List>
  </Dropdown>
</div>
`;
